### DISCUSSION TITLE
EntitySchema with getters documentation not entirely clear

### DISCUSSION THREAD

**Discussion by stefaanMLB** [mikro-orm/mikro-orm#6127] [Q&A]
The documentation on EntitySchema is not entitely clear on how to implement getters.
The documentation defines a `User` class and a `properties` kind-of object but doesn't specify how to turn the `User` object into an entity.
Can this work in the same way as with the reflect-metadata example ?

Here's the example from the [website](https://mikro-orm.io/docs/defining-entities)

```typescript
export class User {

  [HiddenProps]?: 'firstName' | 'lastName';

  firstName!: string;
  lastName!: string;

  getFullName() {
    return `${this.firstName} ${this.lastName}`;
  }

  get fullName2() {
    return `${this.firstName} ${this.lastName}`;
  }
}

properties: {
  firstName: { type: String, hidden: true },
  lastName: { type: String, hidden: true },
  fullName: { type: 'method', persist: false, getter: true, getterName: 'getFullName' },
  fullName2: { type: 'method', persist: false, getter: true },
}
```

**Comment by B4nan** [comment]
That example is just a partial, it builds on the examples above it. Just use `class: User` in the schema definition to map to the class.

**Comment by stefaanMLB** [comment]
With below class definitions, I get `error TS2353: Object literal may only specify known properties, and 'getterName' does not exist in type 'EntitySchemaProperty<string, DriverEntity>'.`
How can I get around that ?

```ts
export class Driver {
  id!: string
  name!: string
  type!: DriverType
}

export class DriverEntity extends Driver {
  [HiddenProps]?: '_name'
  firstname!: string
  lastname!: string
  opTournee!: boolean
  get _name() {
    return `${this.firstname} ${this.lastname.slice(0, 1)}`
  }
}

export const DriverSchema = new EntitySchema<DriverEntity>({
  class: DriverEntity,
  tableName: 'mlb_users',
  properties: {
    id: { type: 'string', length: 60, fieldName: 'login', primary: true },
    firstname: { type: 'string', length: 35, fieldName: 'firstname' },
    lastname: { type: 'string', length: 35, fieldName: 'name' },
    opTournee: { type: JnType, fieldName: 'optournee' },
    type: { enum: true, items: () => DriverTypeValues, length: 10, fieldName: 'code' },
    name: { type: 'method', getter: true, persist: false, getterName: '_name' },
  },
})
```

**Reply by B4nan** [reply]
Hmm, that looks like an omission, since it's not a decorator option (its value is always inferred based on the property that is decorated); it's just a typing issue, it should work the way you tried.

**Reply by stefaanMLB** [reply]
Actually, it doesn't.  The build fails.

**Reply by B4nan** [reply]
Well, TS build only prints errors, but still produces the compiled files. And you can get around type issues with a simple `as any` type assertion (or `@ts-ignore` comments).

**Reply by B4nan** [reply]
Gave this a closer look, and I don't think you should be using the `getterName` here, that is for decorating functions, not getters. Neither the `type: 'method'` is correct, that is again for decorating functions, as in:

```ts
@Property()
getFullname() { ... }
```

For getters, you just use `getter: true`, and use `type` for the actual runtime type your getter will produce, which is a string in this case, that's all. You can already do that in the `EntitySchema` definition. Your getter property is called `_name`, not `name`, and that is what you should have in the key. You are not supposed to be putting the name of your backing variable anywhere in the entity definition, it's your getter that works with that, not the ORM.

```ts
_name: { type: 'string', getter: true, persist: false },
```

But it's quite weird to set it up this way, the getter should be called `name` and your backing property should be `_name`, you got it upside down. Actually, you don't even use this `_name` backing property anywhere. 

Try something like this instead:

```ts
export class DriverEntity {
  id!: string;
  type!: DriverType;
  firstname!: string;
  lastname!: string;
  opTournee!: boolean;

  get name() {
    return `${this.firstname} ${this.lastname.slice(0, 1)}`;
  }
}

// also no need for the explicit type hint when you use `class` option
export const DriverSchema = new EntitySchema({
  class: DriverEntity,
  tableName: 'mlb_users',
  properties: {
    id: { type: 'string', length: 60, fieldName: 'login', primary: true },
    firstname: { type: 'string', length: 35, fieldName: 'firstname' },
    lastname: { type: 'string', length: 35, fieldName: 'name' },
    opTournee: { type: JnType, fieldName: 'optournee' },
    type: { enum: true, items: () => DriverTypeValues, length: 10, fieldName: 'code' },
    name: { type: 'string', getter: true, persist: false },
  },
});
```

**Reply by B4nan** [reply]
I've added the `getterName` to public options here aa18e57c20547d0cdbda27cb2dbfb610c9591e92

But again, it's not what you should be using here.

**Reply by stefaanMLB** [reply]
Hi @B4nan, thanks a lot for your help.
After reading your reply and further reading to docs, I realised that what I need are `formulas`.
There I should mention that the documentation at https://mikro-orm.io/docs/defining-entities#formulas lacks the `entity` property that seems to be required when defining a formula is used.  Example: 

```ts
fullAddress: { formula: 'concat(adres, " ", city)', persist: false, entity: 'DoctorSchema' },
```

Is there a way to define formula in typescript using the other entity properties instead of using SQl syntax ?

**Reply by B4nan** [reply]
You need `type`, not `entity`, that is for relations. 

```ts
fullAddress: { formula: 'concat(adres, " ", city)', persist: false, type: 'string' },
```

> Is there a way to define formula in typescript using the other entity properties instead of using SQl syntax ?

Currently not.

**Reply by stefaanMLB** [reply]
That would be helpfull to not have to pollute the business objects with database intrecacies.
